pkg/parser/translate.go

```diff
@@ -419,21 +418,52 @@ func fromUDPIngressV1Alpha1(log logrus.FieldLogger, ingressList []*v1alpha1.UDPI
 			"udpingress_name":      ingress.Name,
 		})
 
-		result.ServiceNameToServices[ingress.Namespace+"."+ingress.Name] = kongstate.Service{
-			Service: kong.Service{
-				Name:     kong.String(ingress.Namespace + "." + ingress.Name),
-				Protocol: kong.String("udp"),
-				Host:     kong.String(ingressSpec.Host),
-				Port:     kong.Int(ingress.Spec.TargetPort),
-			},
-			Routes: []kongstate.Route{
-				{
-					Route: kong.Route{
-						Protocols:    []*string{kong.String("udp")},
-						Destinations: []*kong.CIDRPort{{Port: kong.Int(ingressSpec.ListenPort)}},
-					},
+		for i, rule := range ingressSpec.Rules {
+			// validate the ports and servicenames for the rule
+			if rule.Port <= 0 || rule.Port >= 65536 {
+				log.Errorf("invalid TCPIngress: invalid port: %d", rule.Port)
+				continue
+			}
+			if rule.Backend.ServiceName == "" {
+				log.Errorf("invalid UDPIngress: empty serviceName")
+				continue
+			}
+			if rule.Backend.ServicePort <= 0 || rule.Backend.ServicePort >= 65536 {
+				log.Errorf("invalid UDPIngress: invalid servicePort: %d", rule.Backend.ServicePort)
+				continue
+			}
+
+			// generate the kong Route based on the listen port
+			route := kongstate.Route{
+				Ingress: util.FromK8sObject(ingress),
+				Route: kong.Route{
+					Name:         kong.String(ingress.Namespace + "." + ingress.Name + "." + strconv.Itoa(i)),
```

Comment:
this naming scheme needs to be guaranteed not to overlap with other routes generated from other rules.

It seems that it would cause a conflict with an ingress named the same as the UDPIngress in the same namespace.

Reply:
Indeed, this was copypasta from `TCPIngress` so it would appear we may already have this problem in `TCPIngress` in the wild and may need to sort that out :thinking:

Reply:
419c371d11a699ea85276f06eca8db5fa90f8776 should take care of this from the UDP perspective, but the problem is historical with TCP and doesn't appear to be something we can change due to backwards compatibility issues (e.g. routes would become orphaned after upgrade unless we add a migration strategy)

Reply:
It looks like we closed the relevant issue without fixing this prior: https://github.com/Kong/kubernetes-ingress-controller/issues/834 It appears the intention was to use UUIDs in the future. Since that's not available yet, I'm going to stick with the `udpingress.` prefix for routes created by udpingress to avoid conflicts but otherwise leave tcpingress alone as out of scope.

I'm also going to re-open https://github.com/Kong/kubernetes-ingress-controller/issues/834 as it appears we need to consider this one further.